cmd/tools/genrpcserverinterceptors/main.go

```diff
@@ -0,0 +1,300 @@
+		if rpcT.NumIn() < 2 {
+			continue
+		}
+		requestT := rpcT.In(1) // Assume request is always the second parameter.
+
+		rd := workflowTagGetters(requestT, 0)
+		rd.Type = requestT.String()
+		sd.Requests = append(sd.Requests, rd)
+	}
+
+	fatalIfErr(template.Must(template.New("code").Parse(tmpl)).Execute(w, sd))
+}
+
+//nolint:revive // cognitive complexity 37 (> max enabled 25)
+func workflowTagGetters(requestT reflect.Type, depth int) requestData {
+	rd := requestData{}
+	if depth > maxRequestDepth {
+		return rd
+	}
+
+	switch {
+	case requestT.AssignableTo(executionGetterT):
+		rd.WorkflowIdGetter = "GetExecution().GetWorkflowId()"
+		rd.RunIdGetter = "GetExecution().GetRunId()"
+	case requestT.AssignableTo(workflowExecutionGetterT):
+		rd.WorkflowIdGetter = "GetWorkflowExecution().GetWorkflowId()"
+		rd.RunIdGetter = "GetWorkflowExecution().GetRunId()"
+	case requestT.AssignableTo(taskTokenGetterT):
+		for _, ert := range excludeTaskTokenTypes {
+			if requestT.AssignableTo(ert) {
+				return rd
+			}
+		}
+		rd.TaskTokenGetter = "GetTaskToken()"
+	default:
+		// Might be one of these, both, or neither.
+		if requestT.AssignableTo(workflowIdGetterT) {
+			rd.WorkflowIdGetter = "GetWorkflowId()"
+		}
+		if requestT.AssignableTo(runIdGetterT) {
+			rd.RunIdGetter = "GetRunId()"
+		}
+	}
+
+	for fieldNum := 0; fieldNum < requestT.Elem().NumField(); fieldNum++ {
+		// First wins: of one of these fields is set, stop iterating.
+		if rd.WorkflowIdGetter != "" || rd.RunIdGetter != "" || rd.TaskTokenGetter != "" {
+			break
+		}
+		nestedRequest := requestT.Elem().Field(fieldNum)
+		if nestedRequest.Type.Kind() != reflect.Ptr {
+			continue
+		}
+		if nestedRequest.Type.Elem().Kind() != reflect.Struct {
+			continue
+		}
+		if depth == 0 {
+			hasAllowedSuffix := false
+			for _, suffix := range rootLevelFieldNameSuffixes {
+				if strings.HasSuffix(nestedRequest.Name, suffix) {
+					hasAllowedSuffix = true
+					break
+				}
+			}
+			if !hasAllowedSuffix {
+				continue
+			}
+		}
+
+		rd = workflowTagGetters(nestedRequest.Type, depth+1)
+		if rd.WorkflowIdGetter != "" {
+			rd.WorkflowIdGetter = fmt.Sprintf("Get%s().%s", nestedRequest.Name, rd.WorkflowIdGetter)
+		}
+		if rd.RunIdGetter != "" {
+			rd.RunIdGetter = fmt.Sprintf("Get%s().%s", nestedRequest.Name, rd.RunIdGetter)
+		}
+		if rd.TaskTokenGetter != "" {
+			rd.TaskTokenGetter = fmt.Sprintf("Get%s().%s", nestedRequest.Name, rd.TaskTokenGetter)
+		}
+	}
+	return rd
+}
+
+func callWithFile(generator func(io.Writer, reflect.Type), server reflect.Type, outPath string, licenseText string) {
+	filename := path.Join(outPath, camelCaseToSnakeCase(server.Name())+"_gen.go")
+	w, err := os.Create(filename)
+	fatalIfErr(err)
+	defer func() {
+		fatalIfErr(w.Close())
+	}()
+	_, err = fmt.Fprintf(w, "%s\n// Code generated by cmd/tools/genrpcserverinterceptors. DO NOT EDIT.\n", licenseText)
+	fatalIfErr(err)
+	generator(w, server)
+}
+
+func readLicenseFile(filePath string) string {
```

Comment:
time to make `common/codegen`?

Reply:
I will do this refactoring in separate PR.